pod.go

```diff
@@ -330,6 +330,12 @@ type PodConfig struct {
 	// Annotations keys must be unique strings and must be name-spaced
 	// with e.g. reverse domain notation (org.clearlinux.key).
 	Annotations map[string]string
+
+	// Devices is a linux specific configuration, these are devices that
+	// must be available within the container.
+	// Appears as a direct child of "linux" object in the OCI json spec.
+	// Note this is different from the devices under "resources".
+	Devices []Device
```

Comment:
@sboeuf VFIO devices get assigned to the VM and the driver will be loaded against the VFIO device. 

Within the VM in the case of network interfaces we can isolate them using network namespaces.  For other use cases we need to if mount namespaces may provide the required isolation.

However until recently kubernetes did not support --device equivalent. We will need to track the k8s device plugin proposal to see where the device will land up.

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md

POD level or container level.

Reply:
@mcastelino Yes I agree the devices are assigned through the VM, but the potential issue if we describe the devices through the Pod structure is that we won't be able to add new devices for further containers. In my mind, it's very close from what we have with the hotplug of block devices.
Doing this at container level allows for more flexibility since it can be handled through the Container structure for every new container, without the need to check which devices have been already created previously (to avoid trying to recreate an existing one).

Reply:
I think we could even create an interface Device that would be held into the Container structure. That way, we would have a list of Device in the container structure, with each Device being a specific one (Block, GenericDevice, ...). Indeed, I think we could process those devices with the same basic rules, the implementation being specific to each structure of this interface.

Reply:
Devices are a per Container resource. The device plugin k8s proposal added devices to the CRI `ContainerConfig` structure, see:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.pb.go#L1596-L1615

So we should make it a per `ContainerConfig` field as well, even though technically it will be a per VM resource for us.